netket/graph/lattice.py

```diff
@@ -246,6 +316,119 @@ def atoms_coord(self):
         """
         return self._atoms_coord
 
+    @property
+    def coords(self):
+        """
+        Returns list of coordinates of lattice points
+        """
+
+        return self._coords
+
+    def translation_perm(self):
+        perms = []
+        for vec in self.basis_vectors:
+            perm = []
+            for coord in self._coords:
+                hash_coord = coord.copy() + vec
+                hash_coord = _np.matmul(hash_coord, self._inv_dims) % 1
+                # make sure 1 and 0 are treated the same
+                hash_coord = hash_coord - hash_coord // (1 - cutoff_tol)
+                hash_coord = (
+                    _np.power(10, tol_digits) * _np.around(hash_coord, tol_digits)
+                ).astype(int)
+                hash_coord = hash(hash_coord.tobytes())
+                perm.append(self._hash_positions[hash_coord])
+
+            perms.append(tuple(perm))
+        return tuple(perms)
+
+    def rotation_perm(self, period, axes=[0, 1]):
+        perm = []
+        axes = list(axes)
+        rot_mat = _np.zeros([2, 2])
+        rot_mat[0, 0] = _np.cos(2 * pi / period)
+        rot_mat[1, 0] = -_np.sin(2 * pi / period)
+        rot_mat[0, 1] = _np.sin(2 * pi / period)
+        rot_mat[1, 1] = _np.cos(2 * pi / period)
+
+        rot_coords = self._coords.copy()
+        rot_coords[:, axes] = _np.matmul(rot_coords[:, axes], rot_mat)
+
+        for hash_coord in rot_coords:
+            hash_coord = _np.matmul(hash_coord, self._inv_dims) % 1
+            hash_coord = hash_coord - hash_coord // (1 - cutoff_tol)
```

Comment:
what does this line do??

Reply:
Apparently float(1) % float(1) sometimes returns 1 and sometimes 0. So I need this extra step to make sure one and zero are treated the same way. Another approach would be to add cutoff_tol to the coordinate before taking the fractional part and hashing

Reply:
But yeah, it would be nice to make this cleaner

Reply:
Ah I see, this is what happens when you don't have IEEE specs for something... 
The more fundamental problem is that we have this "branch cut" between 0 and 1 that should be treated identically. One way around it would be a "circular hash" that hashes cos(2πx) and sin(2πx) rounded to some precision. Because this is a recurring issue in this class, I'll write a general hashing function to do this when I go off to implement #703. I wouldn't worry about it for now, though, if it passes all the nontrivial tests that you provide, it is robust enough to be deployed.

Reply:
There is another thorny issue with the whole hashing thing, though. Imagine that a site is right on the border of two of the hashing bins that we define when rounding 10^5*x to an integer (it's not unreasonable in a large superlattice with a complicated enough unit cell). Then, macheps-level errors can move the site from one bin to the next and cause errors. I'll have a think if there is a hash-based fix for this

Reply:
```suggestion
            hash_coord = hash_coord - hash_coord // 1.0
```
I'd worry that dividing with `1-cutoff_tol` might lead to very slightly negative `hash_coord`s if it was just below 1

Reply:
> There is another thorny issue with the whole hashing thing, though. Imagine that a site is right on the border of two of the hashing bins that we define when rounding 10^5*x to an integer (it's not unreasonable in a large superlattice with a complicated enough unit cell). Then, macheps-level errors can move the site from one bin to the next and cause errors. I'll have a think if there is a hash-based fix for this

~One dumb but working solution would be to store two sets of hashes, one with twice as large bins as the other. Then, if something falls on the edge of a fine bin, it would be guaranteed to be inside (at a quarter-point) of a coarser one.~ This still fails if your lattice point happens to be at the edges of different-width bins in different coordinates.

Another dumb solution (much in the spirit of hashing) would be using 1/some-large-prime-number as the bin width, so getting the outcome I describe above becomes very unlikely. This just feels ugly though :(

@gcarleo, do you think having a transparent O(N^2) check would be too bad? All-to-all distances can be calculated in a vectorised fashion, so we would be saved from long, slow `for` loops, but memory might still be a concern for large lattices (although, we then go on to use GCNNs that convolve with bigger matrices than these...)

Reply:
I would avoid Nˆ2 operations...

Reply:
I tried that fix and it also failed. What if we changed cutoff tol to 10^(-8). Would that make everyone more comfortable?